File: src/main/java/org/athenian/examples/helloworld/GreeterImpl.java
package org.athenian.examples.helloworld;

import com.google.common.collect.Lists;
import io.grpc.stub.StreamObserver;

import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.IntStream;

class GreeterImpl
    extends GreeterGrpc.GreeterImplBase {

  private static final Logger logger = Logger.getLogger(GreeterImpl.class.getName());

  @Override
  public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
    HelloReply reply = HelloReply.newBuilder()
                                 .setMessage("Hello " + request.getName())
                                 .build();
    responseObserver.onNext(reply);
    responseObserver.onCompleted();
  }

  @Override
  public StreamObserver<HelloRequest> sayHelloWithManyRequests(StreamObserver<HelloReply> responseObserver) {
    return
        new StreamObserver<HelloRequest>() {
          final List<String> names = Lists.newArrayList();

          @Override
          public void onNext(HelloRequest request) {
            names.add(request.getName());
          }

          @Override
          public void onError(Throwable t) {
            logger.log(Level.WARNING, "Encountered error in sayHelloWithManyRequests3", t);
          }

          @Override
          public void onCompleted() {
            responseObserver.onNext(HelloReply.newBuilder()
                                              .setMessage("Hello " + String.join(", ", names))
                                              .build());
            responseObserver.onCompleted();
          }
        };
  }

  @Override
  public void sayHelloWithManyReplies(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
    IntStream.range(0, 10)
             .forEach((i) -> {
               HelloReply reply = HelloReply.newBuilder()
                                            .setMessage(String.format("Hello %s [%d]", request.getName(), i))
                                            .build();
               responseObserver.onNext(reply);
             });

    responseObserver.onCompleted();
  }

  @Override
  public StreamObserver<HelloRequest> sayHelloWithManyRequestsAndReplies(StreamObserver<HelloReply> responseObserver) {
    return
        new StreamObserver<HelloRequest>() {
          @Override
          public void onNext(HelloRequest request) {
            IntStream.range(0, 5)
                     .forEach((i) -> {
                       HelloReply reply = HelloReply.newBuilder()
                                                    .setMessage(String.format("Hello %s [%d]", request.getName(), i))
                                                    .build();
                       responseObserver.onNext(reply);
                     });
          }

          @Override
          public void onError(Throwable t) {
            logger.log(Level.WARNING, "Encountered error in sayHelloWithManyRequestsAndReplies", t);
          }

          @Override
          public void onCompleted() {
            responseObserver.onCompleted();
          }
        };
  }
}

File: src/main/kotlin/org/athenian/kotlin_helloworld/withoutCR/GreeterImpl.kt
package org.athenian.kotlin_helloworld.withoutCR

import io.grpc.stub.StreamObserver
import org.athenian.helloworld.GreeterGrpc
import org.athenian.helloworld.HelloReply
import org.athenian.helloworld.HelloRequest

class GreeterImpl : GreeterGrpc.GreeterImplBase() {

    override fun sayHello(request: HelloRequest, responseObserver: StreamObserver<HelloReply>) =
        with(responseObserver) {
            val reply =
                HelloReply.newBuilder()
                    .run {
                        message = "Hello ${request.name}"
                        build()
                    }
            onNext(reply)
            onCompleted()
        }

    override fun sayHelloWithManyRequests(responseObserver: StreamObserver<HelloReply>) =
        object : StreamObserver<HelloRequest> {
            val names: MutableList<String> = mutableListOf()

            override fun onNext(request: HelloRequest) {
                names.add(request.name)
            }

            override fun onError(t: Throwable) {
                println("Encountered error in sayHelloWithManyRequests()")
                t.printStackTrace()
            }

            override fun onCompleted() =
                with(responseObserver) {
                    val msg =
                        HelloReply.newBuilder()
                            .run {
                                message = "Hello ${names.joinToString(", ")}"
                                build()
                            }
                    onNext(msg)
                    onCompleted()
                }
        }

    override fun sayHelloWithManyReplies(request: HelloRequest, responseObserver: StreamObserver<HelloReply>) {
        repeat(5) {
            val reply =
                HelloReply.newBuilder()
                    .run {
                        message = "Hello ${request.name} [$it]"
                        build()
                    }
            responseObserver.onNext(reply)
        }
        responseObserver.onCompleted()
    }

    override fun sayHelloWithManyRequestsAndReplies(responseObserver: StreamObserver<HelloReply>) =
        object : StreamObserver<HelloRequest> {
            override fun onNext(request: HelloRequest) {
                repeat(5) {
                    val reply =
                        HelloReply.newBuilder()
                            .run {
                                message = "Hello ${request.name} [$it]"
                                build()
                            }
                    responseObserver.onNext(reply)
                }
            }

            override fun onError(t: Throwable) {
                println("Encountered error in sayHelloWithManyRequestsAndReplies()")
                t.printStackTrace()
            }

            override fun onCompleted() {
                responseObserver.onCompleted()
            }
        }
}
